packages/db-schema/src/voting/defs.ts

```diff
@@ -210,3 +104,110 @@ export const rewardsRoots = pgTable(
+ * - `updatedAt` should be bumped on every change.
+ * - Rows are mutable — this is the live ledger state.
+ *
+ * Coupling:
+ * - Every update to `boost_balances` must have a matching immutable entry
+ *   in `boost_changes` (same transaction) for auditability/idempotency.
+ *
+ * Typical queries:
+ * - Get balance by wallet and competitionId.
+ */
+export const boostBalances = pgTable(
+  "boost_balances",
+  {
+    id: uuid("id").defaultRandom().primaryKey(),
+    wallet: bytea("wallet").notNull(),
+    competitionId: uuid("competition_id")
+      .notNull()
+      .references(() => competitions.id, { onDelete: "cascade" }),
+    balance: bigint("balance", { mode: "bigint" })
+      .notNull()
+      .default(sql`0`),
+    updatedAt: timestamp("updated_at").notNull().defaultNow(),
+    createdAt: timestamp("created_at").notNull().defaultNow(),
+  },
+  (t) => ({
+    // enforce one balance per (wallet, competition)
+    walletCompetitionUniq: uniqueIndex(
+      "boost_balances_wallet_competition_uniq",
+    ).on(t.wallet, t.competitionId),
+    // address must be exactly 20 bytes
+    walletLenChk: sql`CHECK (octet_length(${t.wallet}) = 20)`,
+    // balance must never be negative
+    balanceNonNegative: sql`CHECK (${t.balance} >= 0)`,
+    balanceDescIdx: index("boost_balances_balance_desc_idx").on(
+      t.competitionId,
+      sql`${t.balance} DESC`,
+    ),
+    updatedIdx: index("boost_balances_updated_at_idx").on(t.updatedAt),
+    createdIdx: index("boost_balances_created_at_idx").on(t.createdAt),
+  }),
+);
+
+/**
+ * boost_changes
+ *
+ * Immutable journal of all Boost mutations.
+ *
+ * Purpose:
+ * - Append-only log of “earn” (+X) and “spend” (−X) operations.
+ * - Provides audit trail, idempotency, and replay capability.
+ *
+ * Invariants / notes:
+ * - (balance_id, idem_key) is unique → an operation is applied at most once per balance (i.e., wallet x competitionId) entry.
+ * - `delta_amount` is signed: positive for earn, negative for spend.
+ * - `meta` holds structured context (competition id, reason, etc).
+ * - Rows are never updated or deleted.
+ *
+ * Coupling:
+ * - Insert into `boost_changes` and update `boost_balances` in the same transaction.
+ * - If the insert is skipped due to the unique constraint, skip the balance mutation
+ *   (idempotent replay).
+ *
+ * Typical queries:
+ * - Fetch wallet history ordered by created_at.
+ * - Check if a given idem_key has already been applied.
+ */
+export const boostChanges = pgTable(
+  "boost_changes",
+  {
+    id: uuid("id").primaryKey().notNull(),
+    balanceId: uuid("balance_id")
+      .notNull()
+      .references(() => boostBalances.id, {
+        onDelete: "restrict",
+        onUpdate: "cascade",
+      }),
+    deltaAmount: bigint("delta_amount", { mode: "bigint" }).notNull(), // earn:+X, spend:-X
+    meta: jsonb("meta")
+      .notNull()
+      .default(sql`'{}'::jsonb`),
+    idemKey: bytea("idem_key").notNull(),
```

Comment:
Any idea how we should create the `idemKey` in practice? I think we do want to support users boosting an agent multiple times in a single competition. Right @jovirecall?

Reply:
`SELECT * FROM table WHERE idemKey = '\xDEADBEEF'::bytea;` Like this.